Optional 2nd package to add radar and battery sensors to esp32-s3-box-3.yaml

jaymunro opened the pull request:
To be added as an extra package in addition to the main esp32-s3-box-3.yaml, e.g.
```
packages:
  esphome.voice-assistant: github://esphome/firmware/wake-word-voice-assistant/esp32-s3-box-3.yaml
  esphome.voice-assistant-sensor: github://jaymunro/esphome_firmware/wake-word-voice-assistant/esp32-s3-box-3-sensor.yaml@sensor_dock
```
All additions only use GPIO pins. I2C bus is not touched so it is very low resource usage.

Provides
---------

Sensors:
 - battery voltage (hidden)
 - battery level %

Binary Sensor:
 - Presence detect (occupancy)

Number:
 - Presence duration (timeout to keep Presence detect true, default 60 seconds)

Switch:
 - Mute when absent (microphone and backlight off for low power)

jesserockz commented:
I like this.

I think we should move it out of this folder though as it can just be used generically for any s3-box-3 and doesn't need to be for voice-assistants. 

Jesse

jaymunro commented:
I and others have had this running for several weeks now with some battery calibration performed from 2 separate users.
See https://community.home-assistant.io/t/esp32-s3-box3/638287/146.

I think having it as a package will make it an easy optional addition for those using the sensor dock.
Not sure on the best location for it though.

jaymunro commented:
> I like this.
> 
> I think we should move it out of this folder though as it can just be used generically for any s3-box-3 and doesn't need to be for voice-assistants.

Ah ok. I'll move it to the voice-assistant folder then...

jaymunro commented:
Development recorded here:
https://github.com/esphome/feature-requests/issues/2475#issuecomment-1879449021

jaymunro commented:
Screenshot of the device with the additional sensors and configuration options
<img width="1013" alt="Screenshot 2024-02-22 at 5 40 13 PM" src="https://github.com/esphome/firmware/assets/1166024/59ecb78a-42f2-44c4-bb9b-6340149b740b">

jaymunro commented:
Presently testing the compile with the yaml:
``` yaml
substitutions:
  name: esp32-s3-box-3-5ad0fc
  friendly_name: Jarvis
  micro_wake_word_model: hey_jarvis
packages:
  esphome.voice-assistant: github://esphome/firmware/wake-word-voice-assistant/esp32-s3-box-3.yaml
  esphome.voice-assistant-sensor: github://jaymunro/esphome_firmware/wake-word-voice-assistant/esp32-s3-box-3-sensor.yaml@sensor_dock
esphome:
  name: ${name}
  name_add_mac_suffix: false
  friendly_name: ${friendly_name}
api:
  encryption:
    key: xxxxxxxxxxxxxxxxxxxxxx=

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
```

jesserockz commented:
Ah yes good point.

I think all the actual sensors could go in one file and mute on a sense could just be documented somewhere that people can add to their adopted yaml even alongside the package line they are already adding

jaymunro commented:
I've been looking through the file structure and it seems the *only* location for box3 firmware is inside the voice-assistant folders. As these sensors only work with the `sensor` version of the dock on the box3 (not even the gen1 box), then there would need to be yet another esp32-s3-box3 folder that is not VA related.

TLDNR: Presently there is no generic location that I can see. Is there presently a non VA use for the Box3 **with sensor dock** in ESPHome?

jaymunro commented:
Perhaps, what could be done is the basic functionality could be modified to be a stand-alone package and go into `esp-web-tools` as `esp32s3box3sensor`. Is this what you had in mind @jesserockz?

How then could this be made obvious to people setting up a box3 with sensor dock as a VA? Can an alias/link/shortcut to that file be put in the voice-assistant folder? Probably not best though imo as additions to it in the future that would work stand alone without VA, may conflict with VA. I am thinking of the i2c bus for temp and humidity as a very good example (there is a serious conflict with VA).

The resolution then may be just to have two versions, one in web tools, the other in voice-assistant (and maybe another in wake-word-voice-assistant?? So that people going for the micro wake word know that the sensor dock components will be compatible).

I am adding lines at the start to make it stand alone for web tools. I can also add the i2c bus with temp & humidity to it as they work and without VA present, VA will not self destruct.

Thoughts?

jaymunro commented:
My further thoughts are that adding a generic firmware for the box3 sensor dock could be useful for someone, I actually put the package together for VA based on a number of forum comments requesting the radar sensor so the box:
- can be carried around as a portable VA and people have an idea when the battery is about to go out.
- goes offline and stops listening/streaming when no-one is in the room (the screen is bright at night.
- when on battery the VA consumes much less power when no-one is in the room.

Can we think about adding it as it is so that people waiting for it can use it now? The addition to web-tools sounds like a whole new project as there are people wanting to add TH and screen interaction that conflicts with VA.

jaymunro commented:
I'm unclear as to what the difference/purpose of the `esp-web-tools` folder vs. the `esphome-web`, so I have added the sensors as a separate generic yaml in `esp-web-tools`.

I have also added comments to the voice-assistant version in the `voice-assistant` folder on how to use it and it's purpose.

balloob commented:
I would suggest that we create a new firmware folder. Each folder has a purpose and we don't want to mix it up.

jaymunro commented:
> I would suggest that we create a new firmware folder. Each folder has a purpose and we don't want to mix it up.

Can you clarify Paulus? Do you mean for the generic firmware or the voice-assistant specific one? Or are you suggesting putting both in a folder like `box3-docks`?

jaymunro commented:
@balloob @jesserockz , I have created a new folder `esp32-box-docks` that could be used for all the Espressif box docks and their various uses.
I've populated it with 2 yaml files:
1. a generic version for the Box3 Sensor dock (battery level & radar for presence/absence)
2. a voice-assistant version for the Box3 Sensor dock (battery & radar with mute on absence option)

If that makes better sense, it should be ready to go.

jesserockz commented on esp32-box-docks/box3-sensor-generic.yaml:
```diff
@@ -0,0 +1,116 @@
+captive_portal:
+
+dashboard_import:
+  package_import_url: github://esphome/firmware/esp-web-tools/esp32s3.yaml@v2
```
This needs to match the path of this file
```suggestion
  package_import_url: github://esphome/firmware/esp32-box-docks/box3-ssensor-generic.yaml@main
```

jaymunro replied:
typo corrected on `ssensor`

jesserockz commented on esp32-box-docks/box3-sensor-generic.yaml:
```diff
@@ -0,0 +1,116 @@
+wifi:
+  # Set up a wifi access point
+  ap:
+    ssid: "ESP Web Tools"
```
```suggestion
```

jaymunro replied:
removed

jesserockz commented on esp32-box-docks/box3-sensor-generic.yaml:
```diff
@@ -0,0 +1,116 @@
+  platformio_options:
+    board_build.flash_mode: dio
+  project:
+    name: esphome.esp_web_tools_example
```
This needs to be updated

jaymunro replied:
used a name of `esphome.esp32-s3-box-3-sensor_example`

jaymunro commented:
@jesserockz, did my last changes meet what you were asking for?

joelis10 commented:
Hey @jaymunro - I'm trying out your custom YAML and so far so good with everything except the battery sensors, I get 
`collect2: error: ld returned 1 exit status`
`*** [.pioenvs/esp32-s3-box-3-5a92ac/firmware.elf] Error 1`
when trying to compile with those sensors added. Commenting them out allows it to compile again but I recently got a battery and would love it to appear in HA too. Any idea what the issue could be?

adrianipopescu commented:
thanks man
